File: app/assets/javascripts/discover/QuoteService.js
(function () {
    var app = angular.module('shocktrade');

    /**
     * Quote Services
     * @author lawrence.daniels@gmail.com
     */
    app.factory('QuoteService', function ($rootScope, $http, $log) {
        var quotes = {};
        var tradingHistory = {};
        var service = {
            lastSymbol: "GOOG"
        };

        function setFavorites(quotes) {
            for (var n = 0; n < quotes.length; n++) {
                quotes[n].favorite = $rootScope.FavoriteSymbols.isFavorite(quotes[n].symbol)
            }
            return quotes
        }

        service.autoCompleteSymbols = function (searchTerm, maxResults) {
            return $http.get('/api/quotes/autocomplete', {params: {'searchTerm': searchTerm, 'maxResults': maxResults}})
        };

        service.loadFilterQuotes = function (filter) {
            if ($rootScope.MySession.isAuthorized()) {
                var id = $rootScope.MySession.getUserID();
                return $http({
                    method: 'POST',
                    url: '/api/profile/' + id + '/quotes/filter/mini',
                    data: angular.toJson(filter)
                })
                    .then(function (response) {
                        return setFavorites(response.data)
                    })
            }
            else {
                return $http({method: 'POST', url: '/api/quotes/filter/mini', data: angular.toJson(filter)})
                    .then(function (response) {
                        return setFavorites(response.data)
                    })
            }
        };

        service.loadStockQuoteList = function (symbols) {
            return $http({method: 'POST', url: '/api/quotes/list', data: angular.toJson(symbols)})
                .then(function (response) {
                    return setFavorites(response.data)
                });
        };

        service.loadStockQuote = function (symbol) {
            return $http.get('/api/quotes/symbol/' + symbol)
        };

        service.loadTradingHistory = function (symbol) {
            return $http.get('/api/quotes/tradingHistory/' + symbol)
                .then(function (response) {
                    return response.data
                })
        };

        service.getRiskLevel = function (symbol) {
            return $http.get('/api/quotes/riskLevel/' + symbol)
        };

        ////////////////////////////////////////////////////////////////////
        //			Exchange Functions
        ///////////////////////////////////////////////////////////////////

        service.getExchangeCounts = function () {
            return $http.get("/api/exchanges")
        };

        service.setExchangeState = function (id, exchange, state) {
            return state
                ? $http.put("/api/profile/" + id + "/exchange/" + exchange)
                : $http.delete("/api/profile/" + id + "/exchange/" + exchange)
        };

        ////////////////////////////////////////////////////////////////////
        //			Sector Exploration Functions
        ///////////////////////////////////////////////////////////////////

        service.loadSectorInfo = function (symbol) {
            return $http.get("/api/explore/symbol/" + symbol)
        };

        service.loadSectors = function () {
            if ($rootScope.MySession.isAuthorized()) {
                var id = $rootScope.MySession.getUserID();
                return $http.get("/api/profile/" + id + "/explore/sectors")
            }
            else {
                return $http.get("/api/explore/sectors")
            }
        };

        service.loadNAICSSectors = function () {
            if ($rootScope.MySession.isAuthorized()) {
                var id = $rootScope.MySession.getUserID();
                return $http.get("/api/profile/" + id + "/explore/naics/sectors")
            }
            else {
                return $http.get("/api/explore/naics/sectors")
            }
        };

        service.loadIndustries = function (sector) {
            if ($rootScope.MySession.isAuthorized()) {
                var id = $rootScope.MySession.getUserID();
                return $http.get("/api/profile/" + id + "/explore/industries", {params: {"sector": sector}})
            }
            else {
                return $http.get("/api/explore/industries", {params: {"sector": sector}})
            }
        };

        service.loadSubIndustries = function (sector, industry) {
            if ($rootScope.MySession.isAuthorized()) {
                var id = $rootScope.MySession.getUserID();
                return $http.get("/api/profile/" + id + "/explore/subIndustries", {
                    params: {
                        "sector": sector,
                        "industry": industry
                    }
                })
            }
            else {
                return $http.get("/api/explore/subIndustries", {params: {"sector": sector, "industry": industry}})
            }
        };

        service.loadIndustryQuotes = function (sector, industry, subIndustry) {
            if ($rootScope.MySession.isAuthorized()) {
                var id = $rootScope.MySession.getUserID();
                return $http.get("/api/profile/" + id + "/explore/quotes", {
                    params: {
                        "sector": sector,
                        "industry": industry,
                        "subIndustry": subIndustry
                    }
                })
            }
            else {
                return $http.get("/api/explore/quotes", {
                    params: {
                        "sector": sector,
                        "industry": industry,
                        "subIndustry": subIndustry
                    }
                })
            }
        };

        ////////////////////////////////////////////////////////////////////
        //			Caching Functions
        ///////////////////////////////////////////////////////////////////

        service.getFilterQuotes = function (filter) {
            return service.loadFilterQuotes(filter)
        };

        service.getPricing = function (symbols) {
            return $http({
                method: 'POST',
                url: '/api/quotes/pricing',
                data: angular.toJson(symbols)
            })
        };

        service.getStockQuote = function (symbol) {
            return service.loadStockQuote(symbol);
        };

        service.getStockQuoteList = function (symbols) {
            return service.loadStockQuoteList(symbols)
        };

        service.getTradingHistory = function (symbol) {
            if (tradingHistory[symbol] != null) return tradingHistory[symbol];
            else {
                var promise = service.loadTradingHistory(symbol);
                tradingHistory[symbol] = promise;
                return promise
            }
        };

        return service
    });

})();

File: app-js/src/main/scala/com/shocktrade/javascript/discover/QuoteService.scala
package com.shocktrade.javascript.discover

import com.github.ldaniels528.meansjs.angularjs.Service
import com.github.ldaniels528.meansjs.angularjs.http.{Http, HttpConfig}
import com.shocktrade.javascript.models.{BSONObjectID, HistoricalQuote, OrderQuote}

import scala.scalajs.js

/**
  * Quote Services
  * @author lawrence.daniels@gmail.com
  */
class QuoteService($http: Http) extends Service {

  def autoCompleteSymbols(searchTerm: String, maxResults: Int) = {
    $http[js.Array[AutoCompletedQuote]](HttpConfig(
      method = "GET",
      url = "/api/quotes/autoComplete",
      params = js.Dictionary("searchTerm" -> searchTerm, "maxResults" -> maxResults)
    ))
  }

  def getExchangeCounts = $http.get[js.Array[js.Dynamic]]("/api/exchanges")

  def getFilterQuotes(aFilter: js.UndefOr[js.Any]) = aFilter foreach { filter =>
    $http.post[js.Dynamic]("/api/quotes/filter/mini", filter)
  }

  def getPricing(symbols: js.Array[String]) = {
    $http.post[js.Dynamic]("/api/quotes/pricing", symbols)
  }

  def getStockQuoteList(symbols: js.Array[String]) = {
    $http.post[js.Array[BasicQuote]]("/api/quotes/list", symbols)
  }

  def getStockQuote(symbol: String) = {
    $http.get[OrderQuote](s"/api/quotes/symbol/$symbol")
  }

  def getTradingHistory(symbol: String) = {
    $http.get[js.Array[HistoricalQuote]](s"/api/quotes/tradingHistory/$symbol")
  }

}

/**
  * Auto-Completed Quote
  * @author lawrence.daniels@gmail.com
  */
@js.native
trait AutoCompletedQuote extends js.Object {
  var _id: js.UndefOr[BSONObjectID] = js.native
  var symbol: js.UndefOr[String] = js.native
  var name: js.UndefOr[String] = js.native
  var exchange: js.UndefOr[String] = js.native
  var assetType: js.UndefOr[String] = js.native
  var icon: js.UndefOr[String] = js.native
}

/**
  * Basic Quote
  * @author lawrence.daniels@gmail.com
  */
@js.native
trait BasicQuote extends js.Object {
  var _id: js.UndefOr[BSONObjectID] = js.native
  var symbol: js.UndefOr[String] = js.native
  var exchange: js.UndefOr[String] = js.native
}
